services/api/service.go

```diff
@@ -493,6 +527,84 @@ func (api *RelayAPI) startValidatorRegistrationDBProcessor() {
 	}
 }
 
+// simulateBlock sends a request for a block simulation to blockSimRateLimiter.
+func (api *RelayAPI) simulateBlock(ctx context.Context, opts blockSimOptions) (error, error) {
+	t := time.Now()
+	reqErr, simErr := api.blockSimRateLimiter.Send(ctx, opts.req, opts.isHighPrio, opts.fastTrack)
+	log := opts.log.WithFields(logrus.Fields{
+		"duration":   time.Since(t).Seconds(),
+		"numWaiting": api.blockSimRateLimiter.CurrentCounter(),
+	})
+	if simErr != nil {
+		ignoreError := simErr.Error() == ErrBlockAlreadyKnown || simErr.Error() == ErrBlockRequiresReorg || strings.Contains(simErr.Error(), ErrMissingTrieNode)
+		if !ignoreError {
+			// Mark builder as non-optimistic.
+			opts.builder.status.IsOptimistic = false
```

Comment:
what's the effect of this? `opts` doesn't seem to be used further down :thinking:

Reply:
hmmm i see, this is a bit non-obvious, `opts` is also passed in as value. 

i'm not feeling great about this function having side-effects.

would it maybe be cleaner if the handling of the errors happens outside of `simulateBlock`?

Reply:
sure, thats fine! we can easily move it to `processOptimisticBlock`. it is probably cleaner there anyways, because we only need to flip the bit when we are in optimistic mode.